capreolus/task/rank.py

```diff
@@ -63,3 +63,23 @@ def evaluate(self):
             logger.info("%25s: %0.4f", metric, score)
 
         return best_results
+
+    def checklist(self):
```

Comment:
This is inspired by [this](https://arxiv.org/abs/2005.04118) paper.
Also, I'm abusing the term "CheckList" to mean the methods used in the ^ paper rather than an actual checklist.

According to our initial discussion, we were trying to discover checklist-worthy samples from each dataset. That would be an issue because:
- We will have to be careful about test data leakage - a sample from train fold can end up in the checklist (since the checklist is handcrafted and static)
- If we create a different CheckList per dataset, the method can't be used for comparing models across datasets (eg: IR engine x trained on robust is better than y trained on MSMarco)
- Underspecification. If I understand correctly, we can say that a model is underspecified if for different random.seeds, we get the same accuracy on the held-out dataset but with a lot of variance in the generalization capability. Hence the underspecification test should be done on a dataset following a different distribution than the training data - our CheckList dataset can be used for this.

I am now leaning towards creating a small set of query doc triplets from scratch. Around 3-4 (query, posdoc, negdoc) pairs for each of the capabilities listed [here](https://docs.google.com/spreadsheets/d/1G3QTx8fpnTCSCNfZpe5GAqfTRnSyv6LAxO6HnEBfKYM/edit?usp=sharing).
That list is just a dummy - we can iterate and come up with a far more refined set of "capabilities" and tests for them during the course of the thesis.

So assuming that we have somehow created a CheckList dataset, we need a way to evaluate our model/algorithm on the checklist. I think the easiest way is to create a benchmark/collection for the CheckList dataset, create an anserini index for this collection, and pass it to the searcher/reranker. The CheckList is static - so we can get away with hard-coding a lot of things.

Reply:
This is a good approach. I think you're right that trying to use instances from an existing collection would be messy.

Reply:
Great. We can brainstorm and come up with concrete behavioral tests once faiss + sentenceBERT is working.